File: 이것이 취업을 위한 코딩 테스트다/구현/치킨_배달/치킨_배달.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int n, m;
int arr[50][50];
vector<pair<int, int>> chicken;
vector<pair<int, int>> house;

int getSum(vector<pair<int, int>> candidates) {
  int result = 0;

  for (int i = 0; i < house.size(); i++) {
    int hx = house[i].first;
    int hy = house[i].second;
    int temp = 1e9;

    for (int j = 0; j < candidates.size(); j++) {
      int x = candidates[j].first;
      int y = candidates[j].second;
      temp = min(temp, abs(x - hx) + abs(y - hy));
    }

    result += temp;
  }

  return result;
}

int main(void) {
  cin >> n >> m;

  for (int r = 0; r < n; r++) {
    for (int c = 0; c < n; c++) {
      cin >> arr[r][c];
      if (arr[r][c] == 1) house.push_back({r, c});
      else if(arr[r][c] == 2) chicken.push_back({r, c});
    }
  }

  vector<bool> binary(chicken.size());
  fill(binary.end()  -m, binary.end(), true);
  for (int i =0; i < binary.size(); i++) {
    cout << binary[i] << ' ';
  }

  int result = 1e9;

  do {
for (int i =0; i < binary.size(); i++) {
    cout << binary[i] << ' ';
  }
    cout << endl;
    vector<pair<int, int>> now;
    for (int i = 0; i < chicken.size(); i++) {
      if (binary[i]) {
        int cx = chicken[i].first;
        int cy = chicken[i].second;
        now.push_back({cx, cy});
      }
    }

    result = min(result, getSum(now));
  } while(next_permutation(binary.begin(), binary.end()));

  cout << result;
}

File: 이것이 취업을 위한 코딩 테스트다/구현/치킨_배달/치킨_배달.java
import java.util.ArrayList;
import java.util.Scanner;

public class 치킨_배달 {
  public static class Combination {
    int n;
    int r;
    int[] now;
    ArrayList<ArrayList<Position>> result;

    Combination(int n, int r) {
      this.n = n;
      this.r = r;
      this.now = new int[r];
      this.result = new ArrayList<ArrayList<Position>>();
    }

    void combination(ArrayList<Position> arr, int depth, int index, int target) {
      if (depth == r) {
        ArrayList<Position> temp = new ArrayList<>();
        for (int i = 0; i < now.length; i++) {
          temp.add(arr.get(now[i]));
        }
        result.add(temp);
        return;
      }

      if (target == n) return;
      now[index] = target;
      combination(arr, depth + 1, index + 1, target + 1);
      combination(arr, depth, index, target + 1);
    }
  }

  public static class Position {
    int x;
    int y;

    Position(int x, int y) {
      this.x = x;
      this.y = y;
    }
  }

  public static int n, m;
  public static int[][] arr = new int[50][50];
  public static ArrayList<Position> chicken = new ArrayList<>();
  public static ArrayList<Position> house = new ArrayList<>();

  public static int getSum(ArrayList<Position> candidates) {
    int result = 0;
    for (int i = 0; i < house.size(); i++) {
      int hx = house.get(i).x;
      int hy = house.get(i).y;
      int temp = (int)1e9;
      for (int j = 0; j < candidates.size(); j++) {
        int cx = candidates.get(j).x;
        int cy = candidates.get(j).y;
        temp = Math.min(temp, Math.abs(hx - cx) + Math.abs(hy - cy));
      }
      result += temp;
    }
    return result;
  }

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    n = sc.nextInt();
    m = sc.nextInt();

    for (int r = 0; r < n; r++) {
      for (int c = 0; c < n; c++) {
        arr[r][c] = sc.nextInt();
        if (arr[r][c] == 1) house.add(new Position(r, c));
        else if (arr[r][c] == 2) chicken.add(new Position(r, c));
      }
    }

    Combination comb = new Combination(chicken.size(), m);
    comb.combination(chicken, 0, 0, 0);
    ArrayList<ArrayList<Position>> chickenList = comb.result;

    int result = (int) 1e9;
    for (int i = 0; i < chickenList.size(); i++) {
      result  = Math.min(result, getSum(chickenList.get(i)));
    }
    System.out.println(result);
  }
}